src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -461,207 +461,152 @@ protected virtual void CheckProjectCanBeRemoved(ProjectId projectId)
         {
         }
 
-        /// <summary>
-        /// Call this method when a project's assembly name is changed in the host environment.
-        /// </summary>
-        protected internal void OnAssemblyNameChanged(ProjectId projectId, string assemblyName)
+        private void HandleProjectChange(ProjectId projectId, Func<ProjectId, Solution, Solution> getSolutionWithChangedProject)
         {
             using (_serializationLock.DisposableWait())
             {
                 CheckProjectIsInCurrentSolution(projectId);
 
                 var oldSolution = this.CurrentSolution;
-                var newSolution = this.SetCurrentSolution(oldSolution.WithProjectAssemblyName(projectId, assemblyName));
+                var newSolution = this.SetCurrentSolution(getSolutionWithChangedProject(projectId, oldSolution));
 
                 this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectChanged, oldSolution, newSolution, projectId);
             }
         }
 
+        /// <summary>
+        /// Call this method when a project's assembly name is changed in the host environment.
+        /// </summary>
+        protected internal void OnAssemblyNameChanged(ProjectId projectId, string assemblyName)
+        {
+            this.HandleProjectChange(projectId, (id, oldSolution) => oldSolution.WithProjectAssemblyName(id, assemblyName));
+        }
+
         /// <summary>
         /// Call this method when a project's output file path is changed in the host environment.
         /// </summary>
         protected internal void OnOutputFilePathChanged(ProjectId projectId, string outputFilePath)
         {
-            using (_serializationLock.DisposableWait())
-            {
-                CheckProjectIsInCurrentSolution(projectId);
-
-                var oldSolution = this.CurrentSolution;
-                var newSolution = this.SetCurrentSolution(oldSolution.WithProjectOutputFilePath(projectId, outputFilePath));
-
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectChanged, oldSolution, newSolution, projectId);
-            }
+            this.HandleProjectChange(projectId, (id, oldSolution) => oldSolution.WithProjectOutputFilePath(id, outputFilePath));
```

Comment:
It's a bit strange here that the lambda gets the ID passed through and we have both "projectId" and "id" available when they're really the same. Given we're still closing over outputFilePath (so the lambda is allocating a closure no matter what), just simplify it a bit?

Reply:
I figured that would be one fewer field in the closure, but that's doesn't really make much difference I guess. Fixed